nixos/modules/services/security/kanidm.nix

```diff
@@ -879,6 +891,41 @@ in
       environment.RUST_LOG = "info";
     };
 
+    systemd.services.kanidm-generate-self-signed = mkIf cfg.generateSelfSigned {
+      description = "small script to generate a self-signed certificate for kanidm";
+      requiredBy = [ "kanidm.service" ];
+      before = [ "kanidm.service" ];
+      enableStrictShellChecks = true;
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = true;
+      };
+      script =
+        let
+          certtool = "${pkgs.gnutls.bin}/bin/certtool";
+          template_file = pkgs.writeText "kanidm-selfsigned-certtool-template" ''
+            cn = "${cfg.serverSettings.domain}"
+            expiration_days = -1
+            signing_key
+            encryption_key
+            tls_www_server
+          '';
+          inherit (cfg.serverSettings) tls_chain tls_key;
+          esc = lib.escapeShellArg;
+        in
+        ''
+          if [[ ! -f ${esc tls_chain} ]]; then
+            directories=("$(dirname ${esc tls_chain})" "$(dirname ${esc tls_key})")
+            mkdir -p "''${directories[@]}"
+            ${certtool} --generate-privkey --outfile=${esc tls_key} --key-type=rsa --sec-param=high
+            ${certtool} --generate-self-signed --load-privkey=${esc tls_key} --outfile=${esc tls_chain} --template=${template_file}
+            chmod 0500 "''${directories[@]}"
+            chmod 0400 ${esc tls_key} ${esc tls_chain}
+            chown kanidm:kanidm ${esc tls_key} ${esc tls_chain} "''${directories[@]}"
```

Comment:
Does this need to run as root? If not, why don't we just add this as a preStart script to kanidm?

Reply:
I struggled with this, the security restrictions on the kanidm service are annoying. It's possible that I just didn't know the right way to do it, but I couldn't get it to work.

Reply:
Also obviously the kanidm user doesn't have permission to create directories and change permissions, so this would have to be split into a `systemd.tmpfiles` entry to create the dir and then a separate service to generate the cert, and they have to be ordered correctly.

Reply:
It should be able to write to its state directory (/var/lib/kanidm), and it looks like you chown the files to the kanidm user at the end. Do you have a specific need to customize the certificate path? Otherwise I believe this should just work by generating the certificate in preStart

Reply:
If you can make it work that'd be rad. I don't need to customize the certificate path, the way I did it just happened to allow it so I saw no need to restrict it.